File: src/puppy_gait/scripts/emergency_response.py
#!/usr/bin/env python3
"""
紧急联动响应节点
- 跌倒检测 → 机器人导航到跌倒位置 + 语音紧急呼叫
- 煤气泄漏 → 机器人导航到厨房确认 + 语音警报
- 入侵检测 → 机器人导航到入侵位置 + 语音警告
- 低电量 → 自动返回充电桩

联动策略:
  1. 接收紧急事件
  2. 取消当前导航任务
  3. 导航到事件位置（失败则重试）
  4. 语音播报紧急信息
  5. 等待事件解除
"""
import json
import math
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.callback_groups import ReentrantCallbackGroup
from nav2_msgs.action import NavigateToPose
from geometry_msgs.msg import Twist
from std_msgs.msg import String, Bool
from action_msgs.msg import GoalStatus


EMERGENCY_LOCATIONS = {
    'fall':      {'x': 1.0, 'y': -2.0, 'yaw': 0.0, 'name': '客厅'},
    'gas_leak':  {'x': 3.0, 'y': 2.0, 'yaw': 0.0, 'name': '厨房'},
    'intrusion': {'x': 0.0, 'y': -1.0, 'yaw': math.pi / 2, 'name': '门口'},
    'charging':  {'x': 0.0, 'y': -2.0, 'yaw': math.pi, 'name': '充电桩'},
}


class EmergencyResponse(Node):
    def __init__(self):
        super().__init__('emergency_response')

        self.cb_group = ReentrantCallbackGroup()

        self.nav_client = ActionClient(
            self, NavigateToPose, 'navigate_to_pose',
            callback_group=self.cb_group)

        self.current_emergency = None
        self.emergency_active = False
        self.last_fall_state = False
        self.last_intrusion = False
        self.last_low_battery = False
        self.cooldown_timer = None
        self.retry_timer = None
        self.goal_handle = None
        self.pending_location = None
        self.retry_count = 0
        self.max_retries = 5

        self.create_subscription(Bool, '/fall_detected', self.fall_callback, 10)
        self.create_subscription(String, '/security/alert', self.security_callback, 10)
        self.create_subscription(Bool, '/security/intrusion', self.intrusion_callback, 10)
        self.create_subscription(Bool, '/low_battery_alert', self.battery_callback, 10)
        self.create_subscription(String, '/emergency/cancel', self.cancel_callback, 10)

        self.voice_pub = self.create_publisher(String, '/voice/tts', 10)
        self.status_pub = self.create_publisher(String, '/emergency/status', 10)
        self.cmd_vel_pub = self.create_publisher(Twist, '/cmd_vel', 10)

        self.create_timer(5.0, self.publish_status)

        self.get_logger().info('紧急联动响应节点已启动')
        self.get_logger().info('监听: /fall_detected, /security/alert, /security/intrusion, /low_battery_alert')

    def fall_callback(self, msg):
        if msg.data and not self.last_fall_state:
            self.get_logger().warn('!!! 紧急事件: 检测到跌倒 !!!')
            self.trigger_emergency('fall', '检测到主人跌倒！正在前往现场确认')
        self.last_fall_state = msg.data

    def security_callback(self, msg):
        alert_text = msg.data
        if '高级警报' in alert_text and '煤气' in alert_text:
            if not self.emergency_active:
                self.trigger_emergency('gas_leak', f'煤气泄漏警报！{alert_text}，正在前往厨房确认')
        elif '高级警报' in alert_text and ('入侵' in alert_text or '闯入' in alert_text):
            if not self.emergency_active:
                self.trigger_emergency('intrusion', f'入侵警报！{alert_text}，正在前往门口查看')

    def intrusion_callback(self, msg):
        if msg.data and not self.last_intrusion:
            if not self.emergency_active:
                self.trigger_emergency('intrusion', '检测到入侵！正在前往门口查看')
        self.last_intrusion = msg.data

    def battery_callback(self, msg):
        if msg.data and not self.last_low_battery:
            if not self.emergency_active:
                self.trigger_emergency('charging', '电量不足，正在自动返回充电桩')
        self.last_low_battery = msg.data

    def trigger_emergency(self, event_type, voice_msg):
        if self.emergency_active:
            self.get_logger().warn(f'已有紧急事件进行中，忽略新事件: {event_type}')
            return

        self.emergency_active = True
        self.current_emergency = event_type
        self.retry_count = 0
        location = EMERGENCY_LOCATIONS.get(event_type, EMERGENCY_LOCATIONS['fall'])
        self.pending_location = location

        self.get_logger().warn(f'===== 紧急响应启动: {event_type} → {location["name"]} =====')

        self.voice_pub.publish(String(data=voice_msg))

        self.navigate_to(location['x'], location['y'], location['yaw'])

    def navigate_to(self, x, y, yaw):
        if self.retry_timer is not None:
            self.retry_timer.cancel()
            self.retry_timer = None

        if not self.nav_client.wait_for_server(timeout_sec=5.0):
            self.get_logger().error('Nav2 服务器不可用，3秒后重试...')
            self.schedule_retry()
            return

        goal_msg = NavigateToPose.Goal()
        goal_msg.pose.header.stamp = self.get_clock().now().to_msg()
        goal_msg.pose.header.frame_id = 'map'
        goal_msg.pose.pose.position.x = float(x)
        goal_msg.pose.pose.position.y = float(y)
        goal_msg.pose.pose.position.z = 0.0
        goal_msg.pose.pose.orientation.z = math.sin(yaw / 2)
        goal_msg.pose.pose.orientation.w = math.cos(yaw / 2)

        self.get_logger().info(
            f'发送导航目标: ({x:.1f}, {y:.1f}, yaw={yaw:.2f}) '
            f'[尝试 {self.retry_count + 1}/{self.max_retries}]')
        send_future = self.nav_client.send_goal_async(goal_msg)
        send_future.add_done_callback(self.goal_response_callback)

    def schedule_retry(self):
        if self.retry_count >= self.max_retries:
            self.get_logger().error(f'导航重试{self.max_retries}次均失败，放弃紧急导航')
            self.emergency_active = False
            self.current_emergency = None
            self.pending_location = None
            return
        self.retry_count += 1
        self.get_logger().info(f'3秒后重试导航 (第{self.retry_count}次)...')
        self.retry_timer = self.create_timer(3.0, self._do_retry)

    def _do_retry(self):
        if self.retry_timer is not None:
            self.retry_timer.cancel()
            self.retry_timer = None
        if self.pending_location is None or not self.emergency_active:
            return
        loc = self.pending_location
        self.navigate_to(loc['x'], loc['y'], loc['yaw'])

    def goal_response_callback(self, future):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().warn('紧急导航目标被拒绝，可能Nav2未就绪')
            self.schedule_retry()
            return

        self.get_logger().info('紧急导航目标已接受')
        self.goal_handle = goal_handle
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self.result_callback)

    def result_callback(self, future):
        status = future.result().status
        self.goal_handle = None
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info(f'已到达紧急事件位置: {self.current_emergency}')
            self.announce_arrival()
            self.cooldown_timer = self.create_timer(10.0, self.clear_emergency)
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().warn(f'导航中止(状态:{status})，执行到达处理')
            self.announce_arrival()
            self.cooldown_timer = self.create_timer(10.0, self.clear_emergency)
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().info(f'导航已取消: {self.current_emergency}')
            self.emergency_active = False
            self.current_emergency = None
            self.pending_location = None
        else:
            self.get_logger().warn(f'导航失败，状态: {status}')
            self.schedule_retry()

    def announce_arrival(self):
        if self.current_emergency == 'fall':
            self.voice_pub.publish(String(
                data='主人，您还好吗？我已经到达现场，正在为您呼叫帮助'))
        elif self.current_emergency == 'gas_leak':
            self.voice_pub.publish(String(
                data='已到达厨房，请立即检查煤气阀门，打开窗户通风，不要使用明火'))
        elif self.current_emergency == 'intrusion':
            self.voice_pub.publish(String(
                data='已到达门口，未发现异常情况，请主人查看监控确认'))
        elif self.current_emergency == 'charging':
            self.voice_pub.publish(String(
                data='已到达充电桩，开始充电'))

    def clear_emergency(self):
        if self.cooldown_timer:
            self.cooldown_timer.cancel()
            self.cooldown_timer = None
        self.get_logger().info(f'紧急事件已解除: {self.current_emergency}')
        self.voice_pub.publish(String(data='紧急事件已解除，恢复正常巡逻'))
        self.emergency_active = False
        self.current_emergency = None
        self.pending_location = None

    def cancel_callback(self, msg):
        self.get_logger().info(f'手动取消紧急事件: {msg.data}')
        if self.cooldown_timer:
            self.cooldown_timer.cancel()
            self.cooldown_timer = None
        if self.retry_timer:
            self.retry_timer.cancel()
            self.retry_timer = None
        if self.goal_handle is not None:
            self.get_logger().info('取消进行中的导航目标')
            cancel_future = self.goal_handle.cancel_goal_async()
            cancel_future.add_done_callback(self.cancel_done_callback)
            self.goal_handle = None
        self.emergency_active = False
        self.current_emergency = None
        self.pending_location = None
        self.cmd_vel_pub.publish(Twist())

    def cancel_done_callback(self, future):
        try:
            result = future.result()
            if len(result.goals_canceling) > 0:
                self.get_logger().info('导航目标已成功取消')
            else:
                self.get_logger().warn('导航目标取消失败')
        except Exception as e:
            self.get_logger().error(f'取消导航目标异常: {e}')

    def publish_status(self):
        status = {
            'active': self.emergency_active,
            'event': self.current_emergency or 'none',
        }
        self.status_pub.publish(String(data=json.dumps(status, ensure_ascii=False)))


def main():
    rclpy.init(args=['--ros-args', '-p', 'use_sim_time:=true'])
    node = EmergencyResponse()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp_src/puppy_gait/emergency_response.cpp
// 紧急联动响应节点实现
#include "puppy_gait/emergency_response.h"

#include <chrono>
#include <cmath>
#include <sstream>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace puppy_gait {

EmergencyLocation EmergencyResponseNode::getEmergencyLocation(
    const std::string &event_type) {
  // 紧急事件位置定义
  if (event_type == "fall") {
    return {1.0, -2.0, 0.0, "客厅"};
  } else if (event_type == "gas_leak") {
    return {3.0, 2.0, 0.0, "厨房"};
  } else if (event_type == "intrusion") {
    return {0.0, -1.0, M_PI / 2, "门口"};
  } else if (event_type == "charging") {
    return {0.0, -2.0, M_PI, "充电桩"};
  }
  // 默认返回 fall 位置
  return {1.0, -2.0, 0.0, "客厅"};
}

EmergencyResponseNode::EmergencyResponseNode()
    : rclcpp::Node("emergency_response") {
  // 使用 ReentrantCallbackGroup 以允许回调内调用动作
  cb_group_ = this->create_callback_group(
      rclcpp::CallbackGroupType::Reentrant);

  // 导航客户端
  nav_client_ = rclcpp_action::create_client<NavigateToPose>(
      this, "navigate_to_pose", cb_group_);

  // 状态
  emergency_active_ = false;
  last_fall_state_ = false;
  last_intrusion_ = false;
  last_low_battery_ = false;
  retry_count_ = 0;
  max_retries_ = 5;

  // 订阅器（使用回调组）
  rclcpp::SubscriptionOptions sub_opts;
  sub_opts.callback_group = cb_group_;

  fall_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/fall_detected", 10,
      std::bind(&EmergencyResponseNode::fallCallback, this,
                std::placeholders::_1),
      sub_opts);
  security_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/security/alert", 10,
      std::bind(&EmergencyResponseNode::securityCallback, this,
                std::placeholders::_1),
      sub_opts);
  intrusion_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/security/intrusion", 10,
      std::bind(&EmergencyResponseNode::intrusionCallback, this,
                std::placeholders::_1),
      sub_opts);
  battery_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/low_battery_alert", 10,
      std::bind(&EmergencyResponseNode::batteryCallback, this,
                std::placeholders::_1),
      sub_opts);
  cancel_sub_ = this->create_subscription<std_msgs::msg::String>(
      "/emergency/cancel", 10,
      std::bind(&EmergencyResponseNode::cancelCallback, this,
                std::placeholders::_1),
      sub_opts);

  // 发布器
  voice_pub_ = this->create_publisher<std_msgs::msg::String>("/voice/tts", 10);
  status_pub_ =
      this->create_publisher<std_msgs::msg::String>("/emergency/status", 10);
  cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

  // 定时发布状态
  status_timer_ = this->create_wall_timer(
      std::chrono::seconds(5), std::bind(&EmergencyResponseNode::publishStatus, this));

  RCLCPP_INFO(this->get_logger(), "紧急联动响应节点已启动");
  RCLCPP_INFO(this->get_logger(),
              "监听: /fall_detected, /security/alert, /security/intrusion, "
              "/low_battery_alert");
}

void EmergencyResponseNode::fallCallback(
    const std_msgs::msg::Bool::SharedPtr msg) {
  if (msg->data && !last_fall_state_) {
    RCLCPP_WARN(this->get_logger(), "!!! 紧急事件: 检测到跌倒 !!!");
    triggerEmergency("fall", "检测到主人跌倒！正在前往现场确认");
  }
  last_fall_state_ = msg->data;
}

void EmergencyResponseNode::securityCallback(
    const std_msgs::msg::String::SharedPtr msg) {
  const std::string &alert_text = msg->data;
  if (alert_text.find("高级警报") != std::string::npos &&
      alert_text.find("煤气") != std::string::npos) {
    if (!emergency_active_) {
      triggerEmergency("gas_leak",
                       "煤气泄漏警报！" + alert_text + "，正在前往厨房确认");
    }
  } else if (alert_text.find("高级警报") != std::string::npos &&
             (alert_text.find("入侵") != std::string::npos ||
              alert_text.find("闯入") != std::string::npos)) {
    if (!emergency_active_) {
      triggerEmergency("intrusion",
                       "入侵警报！" + alert_text + "，正在前往门口查看");
    }
  }
}

void EmergencyResponseNode::intrusionCallback(
    const std_msgs::msg::Bool::SharedPtr msg) {
  if (msg->data && !last_intrusion_) {
    if (!emergency_active_) {
      triggerEmergency("intrusion", "检测到入侵！正在前往门口查看");
    }
  }
  last_intrusion_ = msg->data;
}

void EmergencyResponseNode::batteryCallback(
    const std_msgs::msg::Bool::SharedPtr msg) {
  if (msg->data && !last_low_battery_) {
    if (!emergency_active_) {
      triggerEmergency("charging", "电量不足，正在自动返回充电桩");
    }
  }
  last_low_battery_ = msg->data;
}

void EmergencyResponseNode::triggerEmergency(
    const std::string &event_type, const std::string &voice_msg) {
  if (emergency_active_) {
    RCLCPP_WARN(this->get_logger(),
                "已有紧急事件进行中，忽略新事件: %s", event_type.c_str());
    return;
  }

  emergency_active_ = true;
  current_emergency_ = event_type;
  retry_count_ = 0;
  EmergencyLocation location = getEmergencyLocation(event_type);
  pending_location_ = location;

  RCLCPP_WARN(this->get_logger(),
              "===== 紧急响应启动: %s → %s =====", event_type.c_str(),
              location.name.c_str());

  publishVoice(voice_msg);
  navigateTo(location.x, location.y, location.yaw);
}

void EmergencyResponseNode::navigateTo(double x, double y, double yaw) {
  if (retry_timer_) {
    retry_timer_->cancel();
    retry_timer_.reset();
  }

  if (!nav_client_->wait_for_action_server(std::chrono::seconds(5))) {
    RCLCPP_ERROR(this->get_logger(), "Nav2 服务器不可用，3秒后重试...");
    scheduleRetry();
    return;
  }

  NavigateToPose::Goal goal_msg;
  goal_msg.pose.header.stamp = this->now();
  goal_msg.pose.header.frame_id = "map";
  goal_msg.pose.pose.position.x = x;
  goal_msg.pose.pose.position.y = y;
  goal_msg.pose.pose.position.z = 0.0;
  goal_msg.pose.pose.orientation.z = std::sin(yaw / 2.0);
  goal_msg.pose.pose.orientation.w = std::cos(yaw / 2.0);

  RCLCPP_INFO(this->get_logger(),
              "发送导航目标: (%.1f, %.1f, yaw=%.2f) [尝试 %d/%d]", x, y, yaw,
              retry_count_ + 1, max_retries_);

  auto send_goal_options =
      rclcpp_action::Client<NavigateToPose>::SendGoalOptions();
  send_goal_options.goal_response_callback =
      std::bind(&EmergencyResponseNode::goalResponseCallback, this,
                std::placeholders::_1);
  send_goal_options.result_callback =
      std::bind(&EmergencyResponseNode::resultCallback, this,
                std::placeholders::_1);

  nav_client_->async_send_goal(goal_msg, send_goal_options);
}

void EmergencyResponseNode::goalResponseCallback(
    std::shared_future<rclcpp_action::ClientGoalHandle<NavigateToPose>::SharedPtr>
        future) {
  // 导航目标响应：与 Python 版一致，拒绝则重试，接受则记录 goal_handle
  auto goal_handle = future.get();
  if (!goal_handle) {
    RCLCPP_WARN(this->get_logger(), "紧急导航目标被拒绝，可能Nav2未就绪");
    scheduleRetry();
    return;
  }
  RCLCPP_INFO(this->get_logger(), "紧急导航目标已接受");
  goal_handle_ = goal_handle;
  // 结果由 result_callback 处理
}

void EmergencyResponseNode::scheduleRetry() {
  if (retry_count_ >= max_retries_) {
    RCLCPP_ERROR(this->get_logger(), "导航重试%d次均失败，放弃紧急导航",
                 max_retries_);
    emergency_active_ = false;
    current_emergency_.clear();
    pending_location_.reset();
    return;
  }
  retry_count_++;
  RCLCPP_INFO(this->get_logger(), "3秒后重试导航 (第%d次)...", retry_count_);
  retry_timer_ = this->create_wall_timer(
      std::chrono::seconds(3),
      std::bind(&EmergencyResponseNode::doRetry, this));
}

void EmergencyResponseNode::doRetry() {
  if (retry_timer_) {
    retry_timer_->cancel();
    retry_timer_.reset();
  }
  if (!pending_location_.has_value() || !emergency_active_) {
    return;
  }
  const auto &loc = pending_location_.value();
  navigateTo(loc.x, loc.y, loc.yaw);
}

void EmergencyResponseNode::resultCallback(
    const rclcpp_action::ClientGoalHandle<NavigateToPose>::WrappedResult &
        result) {
  goal_handle_.reset();
  if (result.code == rclcpp_action::ResultCode::SUCCEEDED) {
    RCLCPP_INFO(this->get_logger(), "已到达紧急事件位置: %s",
                current_emergency_.c_str());
    announceArrival();
    cooldown_timer_ = this->create_wall_timer(
        std::chrono::seconds(10),
        std::bind(&EmergencyResponseNode::clearEmergency, this));
  } else if (result.code == rclcpp_action::ResultCode::ABORTED) {
    RCLCPP_WARN(this->get_logger(), "导航中止，执行到达处理");
    announceArrival();
    cooldown_timer_ = this->create_wall_timer(
        std::chrono::seconds(10),
        std::bind(&EmergencyResponseNode::clearEmergency, this));
  } else if (result.code == rclcpp_action::ResultCode::CANCELED) {
    RCLCPP_INFO(this->get_logger(), "导航已取消: %s",
                current_emergency_.c_str());
    emergency_active_ = false;
    current_emergency_.clear();
    pending_location_.reset();
  } else {
    RCLCPP_WARN(this->get_logger(), "导航失败，状态: %d",
                 static_cast<int>(result.code));
    scheduleRetry();
  }
}

void EmergencyResponseNode::announceArrival() {
  // 到达后语音播报
  if (current_emergency_ == "fall") {
    publishVoice("主人，您还好吗？我已经到达现场，正在为您呼叫帮助");
  } else if (current_emergency_ == "gas_leak") {
    publishVoice("已到达厨房，请立即检查煤气阀门，打开窗户通风，不要使用明火");
  } else if (current_emergency_ == "intrusion") {
    publishVoice("已到达门口，未发现异常情况，请主人查看监控确认");
  } else if (current_emergency_ == "charging") {
    publishVoice("已到达充电桩，开始充电");
  }
}

void EmergencyResponseNode::clearEmergency() {
  if (cooldown_timer_) {
    cooldown_timer_->cancel();
    cooldown_timer_.reset();
  }
  RCLCPP_INFO(this->get_logger(), "紧急事件已解除: %s",
              current_emergency_.c_str());
  publishVoice("紧急事件已解除，恢复正常巡逻");
  emergency_active_ = false;
  current_emergency_.clear();
  pending_location_.reset();
}

void EmergencyResponseNode::cancelCallback(
    const std_msgs::msg::String::SharedPtr msg) {
  RCLCPP_INFO(this->get_logger(), "手动取消紧急事件: %s", msg->data.c_str());
  if (cooldown_timer_) {
    cooldown_timer_->cancel();
    cooldown_timer_.reset();
  }
  if (retry_timer_) {
    retry_timer_->cancel();
    retry_timer_.reset();
  }
  if (goal_handle_) {
    RCLCPP_INFO(this->get_logger(), "取消进行中的导航目标");
    auto cancel_future = nav_client_->async_cancel_goal(goal_handle_);
    (void)cancel_future;
    goal_handle_.reset();
  }
  emergency_active_ = false;
  current_emergency_.clear();
  pending_location_.reset();
  geometry_msgs::msg::Twist stop;
  cmd_vel_pub_->publish(stop);
}

void EmergencyResponseNode::cancelDoneCallback(
    rclcpp_action::Client<NavigateToPose>::CancelResponse::SharedPtr
        cancel_response) {
  // 与 Python 版逻辑一致：检查 goals_canceling 是否非空
  if (cancel_response && !cancel_response->goals_canceling.empty()) {
    RCLCPP_INFO(this->get_logger(), "导航目标已成功取消");
  } else {
    RCLCPP_WARN(this->get_logger(), "导航目标取消失败");
  }
}

void EmergencyResponseNode::publishStatus() {
  // 发布状态 (JSON)
  std::ostringstream oss;
  oss << "{\"active\":" << (emergency_active_ ? "true" : "false")
      << ",\"event\":\""
      << (current_emergency_.empty() ? "none" : current_emergency_)
      << "\"}";
  std_msgs::msg::String status;
  status.data = oss.str();
  status_pub_->publish(status);
}

void EmergencyResponseNode::publishVoice(const std::string &text) {
  std_msgs::msg::String msg;
  msg.data = text;
  voice_pub_->publish(msg);
}

}  // namespace puppy_gait
